File: app/src/main/java/de/devmil/paperlaunch/storage/EntriesDataSource.java
package de.devmil.paperlaunch.storage;

import android.content.Context;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteException;

import java.util.ArrayList;
import java.util.List;

import de.devmil.paperlaunch.model.Folder;
import de.devmil.paperlaunch.model.IEntry;
import de.devmil.paperlaunch.model.Launch;

public class EntriesDataSource {
    private EntriesSQLiteOpenHelper mHelper;
    private SQLiteDatabase mDatabase;
    private EntriesAccess mEntriesAccess;
    private FoldersAccess mFoldersAccess;
    private LaunchesAccess mLaunchesAccess;

    private EntriesDataSource() {
    }

    private static final Object sInstanceLockObject = new Object();
    private static EntriesDataSource sInstance = null;
    public static EntriesDataSource getInstance() {
        synchronized (sInstanceLockObject) {
            if (sInstance == null) {
                sInstance = new EntriesDataSource();
            }
            return sInstance;
        }
    }

    public synchronized void accessData(Context context, ITransactionAction action) {
        boolean opened = mDatabase == null;
        if (opened) {
            open(context);
        }
        TransactionContext transactionContext = new TransactionContext();
        action.execute(transactionContext);
        if (opened) {
            close(transactionContext);
        }
    }

    private class TransactionContext implements ITransactionContext {
        @Override
        public void startTransaction() {
            mDatabase.beginTransaction();
        }

        @Override
        public void commitTransaction() {
            if (mDatabase != null && mDatabase.inTransaction()) {
                mDatabase.setTransactionSuccessful();
                mDatabase.endTransaction();
            }
        }

        @Override
        public void rollbackTransaction() {
            if (mDatabase != null && mDatabase.inTransaction()) {
                mDatabase.endTransaction();
            }
        }

        @Override
        public void clear() {
            startTransaction();
            mHelper.clear(mDatabase);
            commitTransaction();
        }

        @Override
        public Launch createLaunch(long parentFolderId) {
            return createLaunch(parentFolderId, -1);
        }

        @Override
        public Launch createLaunch(long parentFolderId, int orderIndex) {
            //create Entry
            EntryDTO entry = mEntriesAccess.createNew(orderIndex);
            entry.setParentFolderId(parentFolderId);
            //create Launch
            LaunchDTO launch = mLaunchesAccess.createNew();
            //relate them
            entry.setLaunchId(launch.getId());

            mEntriesAccess.update(entry);

            return loadLaunch(launch.getId());
        }

        @Override
        public Folder createFolder(long parentFolderId) {
            return createFolder(parentFolderId, -1);
        }

        @Override
        public Folder createFolder(long parentFolderId, int orderIndex) {
            //create Entry
            EntryDTO entry = mEntriesAccess.createNew(orderIndex);
            entry.setParentFolderId(parentFolderId);
            //create folder
            FolderDTO folder = mFoldersAccess.createNew();
            //relate them
            entry.setFolderId(folder.getId());

            mEntriesAccess.update(entry);

            return loadFolder(folder.getId());
        }

        @Override
        public Launch loadLaunch(long launchId) {
            LaunchDTO launch = mLaunchesAccess.queryLaunch(launchId);
            EntryDTO entry = mEntriesAccess.queryEntryForLaunch(launchId);

            return createLaunchFromDTO(launch, entry);
        }

        @Override
        public List<IEntry> loadRootContent() {
            List<EntryDTO> entryDTOs =  mEntriesAccess.queryAllEntries(-1);

            List<IEntry> entries = new ArrayList<>();
            for(EntryDTO entryDto : entryDTOs) {
                entries.add(loadEntry(entryDto));
            }

            return entries;
        }

        private Launch createLaunchFromDTO(LaunchDTO dto, EntryDTO entryDto) {
            return new Launch(dto, entryDto);
        }

        @Override
        public Folder loadFolder(long folderId) {
            FolderDTO folder = mFoldersAccess.queryFolder(folderId);
            EntryDTO entry = mEntriesAccess.queryEntryForFolder(folderId);

            List<EntryDTO> subEntryDTOs = mEntriesAccess.queryAllEntries(folder.getId());

            return createFolderFromDTO(folder, entry, subEntryDTOs);
        }

        @Override
        public void deleteEntry(long entryId) {
            EntryDTO entryDto = mEntriesAccess.queryEntry(entryId);
            if(entryDto != null) {
                if(entryDto.getFolderId() > 0) {
                    deleteFolderContent(entryDto.getFolderId());
                    mFoldersAccess.delete(entryDto.getFolderId());
                }
                else if(entryDto.getLaunchId() > 0) {
                    mLaunchesAccess.delete(entryDto.getLaunchId());
                }
                mEntriesAccess.delete(entryDto);
            }
        }

        private void deleteFolderContent(long folderId) {
            Folder folder = loadFolder(folderId);
            for(IEntry entry : folder.getSubEntries()) {
                deleteEntry(entry.getEntryId());
            }
        }

        private Folder createFolderFromDTO(FolderDTO dto, EntryDTO entryDto, List<EntryDTO> subEntryDTOs) {
            List<IEntry> subEntries = new ArrayList<>();
            for(EntryDTO subEntryDto : subEntryDTOs) {
                subEntries.add(loadEntry(subEntryDto));
            }

            return new Folder(dto, entryDto, subEntries);
        }

        private IEntry loadEntry(EntryDTO entryDto) {
            if(entryDto.getFolderId() > 0) {
                return loadFolder(entryDto.getFolderId());
            } else if(entryDto.getLaunchId() > 0) {
                return loadLaunch(entryDto.getLaunchId());
            }
            return null;
        }

        @Override
        public void updateLaunchData(Launch launch) {
            LaunchDTO launchDto = launch.getDto();

            mLaunchesAccess.update(launchDto);
        }

        @Override
        public void updateFolderData(Folder folder) {
            updateFolderData(folder.getDto());
        }

        @Override
        public void updateFolderData(FolderDTO folderDto) {
            mFoldersAccess.update(folderDto);
        }

        @Override
        public void updateOrders(Folder folder) {
            updateOrders(folder.getSubEntries());
        }

        @Override
        public void updateOrders(List<IEntry> entries) {
            for(int i=0; i<entries.size(); i++) {
                updateOrder(entries.get(i), i);
            }
        }

        @Override
        public void updateOrder(IEntry entry, int orderIndex) {
            EntryDTO entryDTO = mEntriesAccess.queryEntry(entry.getEntryId());
            entryDTO.setOrderIndex(orderIndex);

            mEntriesAccess.update(entryDTO);
        }
    }

    private void open(Context context) throws SQLiteException {
        mHelper = new EntriesSQLiteOpenHelper(context);
        mDatabase = mHelper.getWritableDatabase();
        mEntriesAccess = new EntriesAccess(mDatabase);
        mFoldersAccess = new FoldersAccess(context, mDatabase);
        mLaunchesAccess = new LaunchesAccess(context, mDatabase);
    }

    private void close(ITransactionContext context) {
        context.rollbackTransaction();
        mDatabase.close();
        mDatabase = null;
        mEntriesAccess = null;
        mFoldersAccess = null;
        mLaunchesAccess = null;
        mHelper.close();
    }
}

File: app/src/main/java/de/devmil/paperlaunch/storage/EntriesDataSource.kt
/*
 * Copyright 2015 Devmil Solutions
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package de.devmil.paperlaunch.storage

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteException
import de.devmil.paperlaunch.model.*

import java.util.ArrayList

class EntriesDataSource private constructor() {
    private var helper: EntriesSQLiteOpenHelper? = null
    private var database: SQLiteDatabase? = null
    private var entriesAccess: EntriesAccess? = null
    private var foldersAccess: FoldersAccess? = null
    private var launchesAccess: LaunchesAccess? = null
    private var contextAccess: IContextAccess? = null;

    @Synchronized fun accessData(context: Context, action: ITransactionAction) {
        val opened = database == null
        if (opened) {
            open(context)
        }
        val transactionContext = TransactionContext(contextAccess!!)
        action.execute(transactionContext)
        if (opened) {
            close(transactionContext)
        }
    }

    private inner class TransactionContext(private val contextAccess: IContextAccess) : ITransactionContext {

        override fun startTransaction() {
            database!!.beginTransaction()
        }

        override fun commitTransaction() {
            if (database != null && database!!.inTransaction()) {
                database!!.setTransactionSuccessful()
                database!!.endTransaction()
            }
        }

        override fun rollbackTransaction() {
            if (database != null && database!!.inTransaction()) {
                database!!.endTransaction()
            }
        }

        override fun clear() {
            startTransaction()
            helper!!.clear(database!!)
            commitTransaction()
        }

        override fun createLaunch(parentFolderId: Long): Launch {
            return createLaunch(parentFolderId, -1)
        }

        override fun createLaunch(parentFolderId: Long, orderIndex: Int): Launch {
            //create Entry
            val entry = entriesAccess!!.createNew(orderIndex)
            entry!!.parentFolderId = parentFolderId
            //create Launch
            val launch = launchesAccess!!.createNew()
            //relate them
            entry.launchId = launch!!.id

            entriesAccess!!.update(entry)

            return loadLaunch(launch.id)
        }

        override fun createFolder(parentFolderId: Long, orderIndex: Int, parentFolderDepth: Int): Folder {
            //create Entry
            val entry = entriesAccess!!.createNew(orderIndex)
            entry!!.parentFolderId = parentFolderId
            //create folder
            val folder = foldersAccess!!.createNew()
            folder!!.depth = parentFolderDepth + 1
            foldersAccess!!.update(folder)
            //relate them
            entry.folderId = folder.id

            entriesAccess!!.update(entry)

            return loadFolder(folder.id)
        }

        override fun loadLaunch(launchId: Long): Launch {
            val launch = launchesAccess!!.queryLaunch(launchId)
            val entry = entriesAccess!!.queryEntryForLaunch(launchId)

            return createLaunchFromDTO(launch!!, entry!!)
        }

        override fun loadRootContent(): List<IEntry> {
            val entryDTOs = entriesAccess!!.queryAllEntries(-1)

            val entries = ArrayList<IEntry>()
            for (entryDto in entryDTOs) {
                entries.add(loadEntry(entryDto)!!)
            }

            return entries
        }

        private fun createLaunchFromDTO(dto: LaunchDTO, entryDto: EntryDTO): Launch {
            return Launch(contextAccess, dto, entryDto)
        }

        override fun loadFolder(folderId: Long): Folder {
            val folder = foldersAccess!!.queryFolder(folderId)
            val entry = entriesAccess!!.queryEntryForFolder(folderId)

            val subEntryDTOs = entriesAccess!!.queryAllEntries(folder!!.id)

            return createFolderFromDTO(folder, entry!!, subEntryDTOs)
        }

        override fun deleteEntry(entryId: Long) {
            val entryDto = entriesAccess!!.queryEntry(entryId)
            if (entryDto != null) {
                if (entryDto.folderId > 0) {
                    deleteFolderContent(entryDto.folderId)
                    foldersAccess!!.delete(entryDto.folderId)
                } else if (entryDto.launchId > 0) {
                    launchesAccess!!.delete(entryDto.launchId)
                }
                entriesAccess!!.delete(entryDto)
            }
        }

        private fun deleteFolderContent(folderId: Long) {
            val folder = loadFolder(folderId)
            for (entry in folder.subEntries.orEmpty()) {
                deleteEntry(entry.entryId)
            }
        }

        private fun createFolderFromDTO(dto: FolderDTO, entryDto: EntryDTO, subEntryDTOs: List<EntryDTO>): Folder {
            val subEntries = ArrayList<IEntry>()
            for (subEntryDto in subEntryDTOs) {
                subEntries.add(loadEntry(subEntryDto)!!)
            }

            return Folder(contextAccess, dto, entryDto, subEntries)
        }

        private fun loadEntry(entryDto: EntryDTO): IEntry? {
            if (entryDto.folderId > 0) {
                return loadFolder(entryDto.folderId)
            } else if (entryDto.launchId > 0) {
                return loadLaunch(entryDto.launchId)
            }
            return null
        }

        override fun updateLaunchData(launch: Launch) {
            val launchDto = launch.dto

            launchesAccess!!.update(launchDto)
        }

        override fun updateFolderData(folder: Folder) {
            updateFolderData(folder.dto)
        }

        override fun updateFolderData(folderDto: FolderDTO) {
            foldersAccess!!.update(folderDto)
        }

        override fun updateOrders(folder: Folder) {
            updateOrders(folder.subEntries.orEmpty())
        }

        override fun updateOrders(entries: List<IEntry>) {
            for (i in entries.indices) {
                updateOrder(entries[i], i)
            }
        }

        override fun updateOrder(entry: IEntry, orderIndex: Int) {
            val entryDTO = entriesAccess!!.queryEntry(entry.entryId)

            if (entryDTO != null) {
                entryDTO.orderIndex = orderIndex.toLong()
                entriesAccess!!.update(entryDTO)
            }
        }
    }

    @Throws(SQLiteException::class)
    private fun open(context: Context) {
        helper = EntriesSQLiteOpenHelper(context)
        database = helper!!.writableDatabase
        entriesAccess = EntriesAccess(database!!)
        foldersAccess = FoldersAccess(context, database!!)
        launchesAccess = LaunchesAccess(context, database!!)
        contextAccess = AndroidContextAccess(context)
    }

    private fun close(context: ITransactionContext) {
        context.rollbackTransaction()
        database!!.close()
        database = null
        entriesAccess = null
        foldersAccess = null
        launchesAccess = null
        helper!!.close()
    }

    companion object {

        private val sInstanceLockObject = Any()
        private var sInstance: EntriesDataSource? = null
        val instance: EntriesDataSource
            get() = synchronized(sInstanceLockObject) {
                if (sInstance == null) {
                    sInstance = EntriesDataSource()
                }
                return sInstance!!
            }
    }
}
